Optimizer options

johanbluecreek opened the pull request:
This PR introduces a new field `optimizer_options` for `Options`.

This new field allows the user to change any of the options available for the optimization of constants. That is the user is able to set any of the options of `Optim.Options` [1] via a `NamedTuple`, `Dict`, or as an instance of `Optim.Options` directly:

```julia
julia> using SymbolicRegression
julia> using Optim: Optim
julia> op1 = Options(optimizer_options=(iterations=16,));
julia> op1.optimizer_options.iterations
16
julia> op2 = Options(optimizer_options=Dict(:x_tol => 1.0e-32));
julia> op2.optimizer_options.x_abstol
1.0e-32
julia> op3 = Options(optimizer_options=Optim.Options(; iterations=12));
julia> op3.optimizer_options.iterations
12
```

The `optimizer_iterations` keyword argument to the constructor is kept for backwards compatibility and convenient shorthand, and can be used in combination with `optimizer_options` (`optimizer_options` takes precedence), e.g.

```julia
julia> op4 = Options(optimizer_iterations=2);
julia> op4.optimizer_options.iterations
2
julia> op5 = Options(optimizer_options=(iterations=16,time_limit=1.0), optimizer_iterations=4);
julia> op5.optimizer_options.iterations
16
julia> op5.optimizer_options.time_limit
1.0
```

All tests in `test/runtests.jl` passes for me on this branch.

This PR also contains amendments to the documentation, including clearer statements on what the default algorithm is, how to change it, and also how to use the new `optimizer_options` field.

The `optimizer_iterations` remains undocumented as the same and more functionality exists using `optimizer_options`, which is demonstrated in the documentation.

[1]
https://julianlsolvers.github.io/Optim.jl/v0.9.3/user/config/#general-options

johanbluecreek commented:
I should also mention that in addition to what was discussed in the #120 issue, I allowed for `NamedTuple` (as well as `Dict`) as it is a bit more convenient.

I did not implement any unit tests for this new feature. Let me know if you want me to write some.

MilesCranmer commented:
To solve the formatting issue, just run `julia -e 'using JuliaFormatter; format(".", BlueStyle())'` in the root folder.

johanbluecreek commented:
Maybe I should also write something so you know I'm done on my side... I added three unit-tests, one for each possible input type. Fixed the errors in formatting using the command you gave (thank you). Then I also resolved the conflicts that had appeared with master.

MilesCranmer reviewed (approved):
Looks great, thanks!
